cpak.1
.TH CPAK 1 cpak-__VERSION__
.SH NAME
cpak - the fast and efficient source based package manager for the c programming language 
.SH SYNOPSIS
.B cpak
.RI [ COMMAND ]
.RI [ SUBCOMMAND | FLAG ]
.P
.SH DESCRIPTION
Cpak makes it easy for the developer to manage packages for c projects,
by allowing devs to store project dependencies globally or inside a project. 
It also enables a templating and bootstraping toolkit. 
Cpak features a extensive documentation and a very verbose logging system 
to warn the users of malformatted config files and several other nuances.

.TP
.B version

Show the program version.

.TP
.B help

Verbose help for every command and subcommand available. 
If no subcommand is given, the cpak usage is displayed.
This command does not accept short version of commands, use full names.

.TP
.B config

Generates the default cpak config and the default template in
.B CPAK_CONFIG_HOME/cpak 
or 
.B XDG_CONFIG_HOME/cpak.
If there is already a configuration present at one of these locations the config will not be overridden this config unless the
.B -f 
force flag is specifed as a subcommand.

.TP
.B bootstrap

Copies the files specified in a subdirectory in 
.B CPAK_CONFIG_HOME/cpak/templates/<template_name>
or 
.B XDG_CONFIG_HOME/cpak/templates/<template_name>.
Also executes the 
.B cpak_template.sh
file found in the template directory in the newly bootstraped directory.

.TP
.B add

Adds a new module to the module registry in 
.B CPAK_MODULE_HOME
or
.B XDG_DATA_HOME/cpak.
This command also creates a soft link in the current directory to the installed module in the
.B cpak_modules
directory. This command requires the module to be formatted in the following way:

<repo_owner>/<repo_name>

.TP
.B remove

Removes the specified module from the module registry in 
.B CPAK_MODULE_HOME
or
.B XDG_DATA_HOME/cpak.
This command also unlinks the module soft link in the current directory to the installed module in the
.B cpak_modules
directory. In contrary to the 
.B add 
command, this command only needs the module name.

.TP
.B upgrade

Syncs the specified installed module to the newest upstream commit.
This command also only needs the module name as a subcommand.

.SH EXAMPLES:
The following examples make use of the short version commands, all of them can be called by the full name of the command. 

.TP
.B version

Get the installed cpak version.
.P
.nf                                                       
.RS                                                       
cpak v
.RE                                                       
.fi    

.TP
.B help

Get information about the help command.
.P
.nf                                                       
.RS                                                       
cpak h help
.RE                                                       
.fi    

.TP
.B config

Generate the default cpak config and the default template.
.P
.nf                                                       
.RS                                                       
cpak c
.RE                                                       
.fi    
.P
.nf                                                       
.RS                                                       
cpak c -f
.RE                                                       
.fi    

.TP
.B bootstrap

Copies the content of the template 
.B default 
into the current directory and executes the 
.B cpak_template.sh
file in the current directory.
.P
.nf                                                       
.RS                                                       
cpak b default
.RE                                                       
.fi    

.TP
.B add

Install the module
.B xutils
from the Github user 
.B xnacly.
.P
.nf                                                       
.RS                                                       
cpak a xnacly/xutils
.RE                                                       
.fi    

.TP
.B remove

Remove the module
.B xutils.
.P
.nf                                                       
.RS                                                       
cpak r xutils
.RE                                                       
.fi    

.TP
.B upgrade

Upgrade the module
.B xutils.
.P
.nf                                                       
.RS                                                       
cpak u xutils
.RE                                                       
.fi    
.SH AUTHORS
Written by xnacly

.SH REPORTING BUGS
Cpak online documentation: <https://github.com/xnacly/c_paket>

Report bugs to: <https://github.com/xnacly/c_paket/issues/new>



src/cpak.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "./core/cpak_cli.h"
#include "./core/cpak_config.h"
#include "./core/cpak_utils.h"
#include "cpak.h"

void help_wrapper() {
    printf("cpak-%s\n", CPAK_VERSION);
    c_help("all");
}

Cli_arguments *parse_arguments(int arguments_amount, char **arguments) { Cli_arguments *ca;
    ca = malloc(sizeof(*ca));
    ca->c_cmd = malloc(sizeof(arguments[2]));

    if (arguments_amount == 1) {
        ca->cmd = UNDEF;
        return ca;
    }
    char *cmd = arguments[1];

    if (s_is_equal(cmd, "help") || s_is_equal(cmd, "h")) {
        ca->cmd = HELP;
    } else if (s_is_equal(cmd, "bootstrap") || s_is_equal(cmd, "b")) {
        ca->cmd = BOOTSTRAP;
    } else if (s_is_equal(cmd, "config") || s_is_equal(cmd, "c")) {
        ca->cmd = CONFIG;
    } else if (s_is_equal(cmd, "init") || s_is_equal(cmd, "i")) {
        ca->cmd = INIT;
    } else if (s_is_equal(cmd, "add") || s_is_equal(cmd, "a")) {
        ca->cmd = ADD;
    } else if (s_is_equal(cmd, "remove") || s_is_equal(cmd, "r")) {
        ca->cmd = REMOVE;
    } else if (s_is_equal(cmd, "upgrade") || s_is_equal(cmd, "u")) {
        ca->cmd = UPGRADE;
    } else if (s_is_equal(cmd, "version") || s_is_equal(cmd, "v")) {
        ca->cmd = VERSION;
    } else {
        ca->cmd = UNKNOWN;
    }

    if (ca->cmd != VERSION) {
        if (arguments_amount >= 2) {
            ca->c_cmd = arguments[2];
        }
    }

    return ca;
}

int main(int argc, char *argv[]) {
    int feedback = EXIT_FAILURE;
    int does_conf_exists = conf_exists();
    Cli_arguments *arg = parse_arguments(argc, argv);
    if (!does_conf_exists && arg->cmd != CONFIG)
        throw_warning("can't find cpak's config file, consider generating the "
                      "default config using: 'cpak config'",
                      ERR_CONF_MISSING_CONFIG);
    get_config();

    switch (arg->cmd) {
        case BOOTSTRAP:
            feedback = c_bootstrap(arg->c_cmd, ".");
            break;
        case INIT:
            throw_warning("this command is not available", ERR_NOT_IMPLEMENTED);
            /* if (!s_is_empty(arg->c_cmd)) { */
            /*     feedback = c_init(arg->c_cmd); */
            /* } else { */
            /*     throw_warning("Not enough Arguments providen", NOT_ENOUGH_ARGUMENTS); */
            /*     c_help("init"); */
            /* } */
            break;
        case CONFIG: {
            int o = 0;
            if (!s_is_empty(arg->c_cmd) && s_is_equal(arg->c_cmd, "-f")) {
                o = 1;
            }
            feedback = c_config(o);
            break;
        }
        case ADD:{
             if(s_is_empty(arg->c_cmd)){
                 throw_error("Please specify a module to install", ERR_NOT_ENOUGH_ARGUMENTS);
             }
            feedback = c_add(arg->c_cmd);
            break;
        }
        case REMOVE:{
             if(s_is_empty(arg->c_cmd)){
                 throw_error("Please specify a module to remove", ERR_NOT_ENOUGH_ARGUMENTS);
             }
            feedback = c_remove(arg->c_cmd);
            break;
        }
        case UPGRADE: {
             if(s_is_empty(arg->c_cmd)){
                 throw_error("Please specify a module to upgrade", ERR_NOT_ENOUGH_ARGUMENTS);
             }
            feedback = c_upgrade(arg->c_cmd);
            break;
        }
        case VERSION:
            cpak_log("cpak_" CPAK_VERSION, INFO);
            break;
        case UNDEF:
            throw_warning("Not enough Arguments providen", ERR_NOT_ENOUGH_ARGUMENTS);
            help_wrapper();
            break;
        case HELP:
            if (!s_is_empty(arg->c_cmd)) {
                c_help(arg->c_cmd);
            } else {
                help_wrapper();
            }
            feedback = EXIT_SUCCESS;
            break;
        case UNKNOWN:
            throw_warning("Unknown Argument", ERR_UNKNOWN_ARGUMENT);
            help_wrapper();
            break;
        default:
            help_wrapper();
            break;
    }

    free(arg);
    return feedback;
}
